File: C/SLLNC.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Struktur untuk merepresentasikan Mahasiswa
struct Mahasiswa {
    char nim[20];
    char nama[50];
    struct Mahasiswa* next;
};

// Struktur untuk merepresentasikan Linked List
struct LinkedList {
    struct Mahasiswa* head;
};

// Fungsi untuk membuat Mahasiswa baru
struct Mahasiswa* createMahasiswa(const char* nim, const char* nama) {
    struct Mahasiswa* newMahasiswa = (struct Mahasiswa*)malloc(sizeof(struct Mahasiswa));
    snprintf(newMahasiswa->nim, sizeof(newMahasiswa->nim), "%s", nim);
    snprintf(newMahasiswa->nama, sizeof(newMahasiswa->nama), "%s", nama);
    newMahasiswa->next = NULL;
    return newMahasiswa;
}

// Fungsi untuk menambahkan Mahasiswa ke Linked List Non-Circular
void tambahMahasiswa(struct LinkedList* list, const char* nim, const char* nama) {
    struct Mahasiswa* newMahasiswa = createMahasiswa(nim, nama);

    if (list->head == NULL) {
        list->head = newMahasiswa;
    } else {
        struct Mahasiswa* temp = list->head;
        while (temp->next != NULL) {
            temp = temp->next;
        }
        temp->next = newMahasiswa;
    }
}

// Fungsi untuk menampilkan semua Mahasiswa dalam Linked List
void tampilkanMahasiswa(struct LinkedList* list) {
    if (list->head == NULL) {
        printf("Linked List Kosong.\n");
        return;
    }

    struct Mahasiswa* temp = list->head;
    while (temp != NULL) {
        printf("NIM: %s, Nama: %s\n", temp->nim, temp->nama);
        temp = temp->next;
    }
}

// Fungsi untuk menghapus Mahasiswa dari Linked List berdasarkan NIM
void hapusMahasiswa(struct LinkedList* list, const char* nim) {
    if (list->head == NULL) {
        printf("Linked List Kosong. Tidak dapat menghapus.\n");
        return;
    }

    struct Mahasiswa* temp = list->head;
    struct Mahasiswa* prev = NULL;

    while (temp != NULL && strcmp(temp->nim, nim) != 0) {
        prev = temp;
        temp = temp->next;
    }

    if (temp == NULL) {
        printf("Mahasiswa dengan NIM %s tidak ditemukan.\n", nim);
        return;
    }

    if (prev == NULL) {
        // Mahasiswa yang dihapus adalah elemen pertama
        list->head = temp->next;
    } else {
        // Mahasiswa yang dihapus berada di tengah atau akhir
        prev->next = temp->next;
    }

    free(temp);
    printf("Mahasiswa dengan NIM %s berhasil dihapus.\n", nim);
}

// Fungsi untuk merubah data mahasiswa secara manual
void ubahData(struct LinkedList* list, const char* nim) {
    struct Mahasiswa* temp = list->head;

    while (temp != NULL && strcmp(temp->nim, nim) != 0) {
        temp = temp->next;
    }

    if (temp == NULL) {
        printf("Mahasiswa dengan NIM %s tidak ditemukan.\n", nim);
        return;
    }

    // Mahasiswa ditemukan, izinkan pengguna untuk merubah nama
    printf("Nama saat ini: %s\n", temp->nama);
    printf("Masukkan nama baru: ");
    scanf(" %[^\n]s", temp->nama);
    printf("Data mahasiswa berhasil diubah.\n");
}

// Fungsi utama
int main() {
    struct LinkedList daftarMahasiswa = {NULL};
    int pilihan;
    char nim[20], nama[50];

    do {
        printf("\nMenu:\n");
        printf("1. Tambah Mahasiswa\n");
        printf("2. Tampilkan Mahasiswa\n");
        printf("3. Hapus Mahasiswa\n");
        printf("4. Ubah Data Mahasiswa\n");
        printf("0. Keluar\n");
        printf("Pilihan Anda: ");
        scanf("%d", &pilihan);

        switch (pilihan) {
            case 1:
                printf("Masukkan NIM mahasiswa: ");
                scanf("%s", nim);

                printf("Masukkan Nama mahasiswa: ");
                scanf(" %[^\n]s", nama);

                tambahMahasiswa(&daftarMahasiswa, nim, nama);
                break;

            case 2:
                printf("\nDaftar Mahasiswa:\n");
                tampilkanMahasiswa(&daftarMahasiswa);
                break;

            case 3:
                printf("Masukkan NIM mahasiswa yang ingin dihapus: ");
                scanf("%s", nim);
                hapusMahasiswa(&daftarMahasiswa, nim);
                break;

            case 4:
                printf("Masukkan NIM mahasiswa yang ingin diubah: ");
                scanf("%s", nim);
                ubahData(&daftarMahasiswa, nim);
                break;

            case 0:
                printf("Program selesai.\n");
                break;

            default:
                printf("Pilihan tidak valid. Silakan coba lagi.\n");
        }
    } while (pilihan != 0);

    // Membersihkan memori sebelum keluar dari program
    struct Mahasiswa* temp = daftarMahasiswa.head;
    while (temp != NULL) {
        struct Mahasiswa* next = temp->next;
        free(temp);
        temp = next;
    }

    return 0;
}

File: C/SLLNC.py
class Node:
    def __init__(self, nim, nama):
        self.nim = nim
        self.nama = nama
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def tambah_mahasiswa(self, nim, nama):
        new_mahasiswa = Node(nim, nama)
        if self.head is None:
            self.head = new_mahasiswa
        else:
            temp = self.head
            while temp.next is not None:
                temp = temp.next
            temp.next = new_mahasiswa

    def tampilkan_mahasiswa(self):
        if self.head is None:
            print("Linked List Kosong.")
        else:
            temp = self.head
            while temp is not None:
                print(f"NIM: {temp.nim}, Nama: {temp.nama}")
                temp = temp.next

    def hapus_mahasiswa(self, nim):
        if self.head is None:
            print("Linked List Kosong.")
            return

        if self.head.nim == nim:
            self.head = self.head.next
            print(f"Mahasiswa dengan NIM {nim} berhasil dihapus.")
            return

        temp = self.head
        while temp.next is not None and temp.next.nim != nim:
            temp = temp.next

        if temp.next is None:
            print(f"Mahasiswa dengan NIM {nim} tidak ditemukan.")
        else:
            temp.next = temp.next.next
            print(f"Mahasiswa dengan NIM {nim} berhasil dihapus.")

    def ubah_data_mahasiswa(self, nim):
        if self.head is None:
            print("Linked List Kosong.")
            return

        temp = self.head
        while temp is not None and temp.nim != nim:
            temp = temp.next

        if temp is None:
            print(f"Mahasiswa dengan NIM {nim} tidak ditemukan.")
        else:
            print(f"Nama saat ini: {temp.nama}")
            new_nama = input("Masukkan nama baru: ")
            temp.nama = new_nama
            print("Data mahasiswa berhasil diubah.")


def main():
    daftar_mahasiswa = LinkedList()

    while True:
        print("\nMenu:")
        print("1. Tambah Mahasiswa")
        print("2. Tampilkan Mahasiswa")
        print("3. Hapus Mahasiswa")
        print("4. Ubah Data Mahasiswa")
        print("0. Keluar")

        pilihan = input("Pilihan Anda: ")

        if pilihan == '1':
            nim = input("Masukkan NIM mahasiswa: ")
            nama = input("Masukkan nama mahasiswa: ")
            daftar_mahasiswa.tambah_mahasiswa(nim, nama)
        elif pilihan == '2':
            print("\nDaftar Mahasiswa:")
            daftar_mahasiswa.tampilkan_mahasiswa()
        elif pilihan == '3':
            nim_hapus = input("Masukkan NIM mahasiswa yang ingin dihapus: ")
            daftar_mahasiswa.hapus_mahasiswa(nim_hapus)
        elif pilihan == '4':
            nim_ubah = input("Masukkan NIM mahasiswa yang ingin diubah: ")
            daftar_mahasiswa.ubah_data_mahasiswa(nim_ubah)
        elif pilihan == '0':
            print("Program selesai.")
            break
        else:
            print("Pilihan tidak valid. Silakan coba lagi.")


if __name__ == "__main__":
    main()
